torch/csrc/api/src/data/ctf/reader.cpp

```diff
@@ -0,0 +1,94 @@
+#include <torch/data/ctf/reader.h>
+#include <torch/data/ctf/reader_constants.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <string>
+
+namespace torch {
+namespace data {
+namespace ctf {
+
+/*
+ * Reader class for CTF
+ *
+ * RAII pattern was used for file descriptor
+ */
+
+Reader::~Reader() {}
+
+Reader::Reader(const std::string& filename)
+    : filename_(filename),
+      file_pos_(0),
+      is_eof_(false),
+      buffer_pos_(0),
+      buffer_size_(0),
+      rewinded_char_(false),
+      previous_char_(0) {
+  std::FILE* const tmp = fopen(filename_.c_str(), "rbS");
+  if (!tmp) {
+    std::string error_msg(
+        "Reader could not open the specified file (" + filename + ")");
+#ifdef CTF_DEBUG
+    std::cerr << error_msg << std::endl;
+#endif
+    throw std::runtime_error(error_msg);
+  }
+  // Get file length
+  file_ = std::shared_ptr<std::FILE>(tmp, std::fclose);
+  std::fseek(file_.get(), 0, SEEK_END); // TODO: Non-portable as binary streams
+                                        // are not required to support SEEK_END
+  file_size_ = std::ftell(file_.get());
+  std::rewind(file_.get());
+
+   buffer_.resize(Reader::CTF_MAX_BUFFER_SIZE);
+  refill();
+}
+
+bool Reader::refill(void) {
+  if (!is_buffer_empty()) {
+#ifdef CTF_DEBUG
+    std::cout << "Buffer is not empty yet. Not refilling it" << std::endl;
+#endif
+    return false;
+  }
+  if (!can_buffer()) {
+#ifdef CTF_DEBUG
+    std::cout << "Nothing to read from file " << filename_ << ". ("
+              << strerror(errno) << ")";
+#endif
+    is_eof_ = true;
+    return false;
+  }
+
+  buffer_pos_ = 0;
+  file_pos_ = ftell(file_.get());
```

Comment:
Similar to the previous comment, could we remember this?

Reply:
file_pos_ not needed anymore